### PR TITLE
Experimental TSDB: Enable Azure Storage Backend

### PR DESCRIPTION
**Pull request by khaines**
**What this PR does**: Adds the needed cortex structs and configuration to enable azure storage usage with the experimental tsdb engine. 

I've got a small cluster using this currently as it's storage backend, but it's still relatively new so I would say testing for this over the long term is still ongoing.

**Checklist**
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storage/tsdb/backend/azure/config.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package azure
+
+import (
+	"flag"
+)
+
+// Config holds the config options for an Azure backend
+type Config struct {
+	StorageAccountName string `yaml:"storage_account"`
```
Comment (pracucci) [inline pkg/storage/tsdb/backend/azure/config.go]: `storage_account` > `account_name` (to keep it consistent with the CLI flag)

File: pkg/storage/tsdb/backend/azure/config.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	StorageAccountKey  string `yaml:"storage_account_key"`
```
Comment (pracucci) [inline pkg/storage/tsdb/backend/azure/config.go]: `storage_account_key` > `account_key` (to keep it consistent with the CLI flag)

File: pkg/storage/tsdb/backend/azure/config.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	ContainerName      string `yaml:"container"`
```
Comment (pracucci) [inline pkg/storage/tsdb/backend/azure/config.go]: `container` > `container_name` (to keep it consistent with the CLI flag)

File: pkg/storage/tsdb/backend/azure/config.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	MaxRetries         int    `yaml:"max_retries"`
+}
+
+// RegisterFlags registers the flags for TSDB Azure storage
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.StorageAccountName, "experimental.tsdb.azure.account-name", "", "Azure storage account name")
+	f.StringVar(&cfg.StorageAccountKey, "experimental.tsdb.azure.account-key", "", "Azure storage account key")
+	f.StringVar(&cfg.ContainerName, "experimental.tsdb.azure.container-name", "", "Azure storage container name")
+	f.StringVar(&cfg.Endpoint, "experimental.tsdb.azure.endpoint-suffix", "", "Azure storage endpoint suffix without schema. The account name will be prefixed to this value to create the FQDN")
+	f.IntVar(&cfg.MaxRetries, "experimental.tsdb.azure.max-retries", 20, "Number of retries for recoverable errors (default 20)")
```
Comment (pracucci) [inline pkg/storage/tsdb/backend/azure/config.go]: Please remove the ` (default 20)` from the flag description. The default is automatically injected with the `--help` or config file doc are generated.

File: docs/operations/blocks-storage.md [hunk 5]
```diff
@@ -11,6 +11,7 @@ The supported backends for the blocks storage are:
 
 * [Amazon S3](https://aws.amazon.com/s3)
 * [Google Cloud Storage](https://cloud.google.com/storage/)
+* [Microsoft Azure Storage](https://azure.microsoft.com/en-us/services/storage/)
```
Comment (pracucci) [inline docs/operations/blocks-storage.md]: May you also reference it in the `docs/architecture.md` within the section "Blocks storage (experimental)", please?

Comment (khaines) [reply docs/operations/blocks-storage.md]: Yes, thank you for calling that out.

File: pkg/storage/tsdb/backend/azure/config.go (hunk 4 above)
Comment (khaines) [reply pkg/storage/tsdb/backend/azure/config.go]: fixed

File: docs/operations/blocks-storage.md [hunk 6]
```diff
@@ -189,6 +189,26 @@ tsdb:
     # Google SDK default logic.
     # CLI flag: -experimental.tsdb.gcs.service-account string
     [ service_account: <string>]
+
+  # Configures the Azure storage backend
+  # Required only when "azure" backend has been selected.
+  azure:
+    # Azure storage account name
+    # CLI flag: -experimental.tsdb.azure.account-name
+    storage_account: <string>
```
Comment (pracucci) [inline docs/operations/blocks-storage.md]: `storage_account` > `account_name`

File: docs/operations/blocks-storage.md [hunk 7]
```diff
@@ -189,6 +189,26 @@ tsdb:
...
+    # Azure storage account key
+    # CLI flag: -experimental.tsdb.azure.account-key
+    storage_account_key: <string>
```
Comment (pracucci) [inline docs/operations/blocks-storage.md]: `storage_account_key ` > `account_key`

File: docs/operations/blocks-storage.md [hunk 8]
```diff
@@ -189,6 +189,26 @@ tsdb:
...
+    # Azure storage container name
+    # CLI flag: -experimental.tsdb.azure.container-name
+    container: <string>
```
Comment (pracucci) [inline docs/operations/blocks-storage.md]: `container` > `container_name`

**Comment by khaines** [comment]
Thanks for the attention to detail on the documentation @pracucci !

File: docs/operations/blocks-storage.md (hunk 8 above)
Comment (khaines) [reply docs/operations/blocks-storage.md]: fixed

File: docs/operations/blocks-storage.md (hunk 7 above)
Comment (khaines) [reply docs/operations/blocks-storage.md]: fixed

File: docs/operations/blocks-storage.md (hunk 6 above)
Comment (khaines) [reply docs/operations/blocks-storage.md]: fixed

**Pull request merged**